server/src/main/scala/services/Optimiser.scala

```diff
@@ -0,0 +1,373 @@
+package services
+
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.mutable
+import scala.util.{Success, Try}
+
+
+case class ProcessedWork(util: List[Double],
+                         waits: List[Int],
+                         residual: IndexedSeq[Double],
+                         totalWait: Double,
+                         excessWait: Double)
+
+case class Cost(paxPenalty: Double, slaPenalty: Double, staffPenalty: Double, churnPenalty: Int, totalPenalty: Double)
+
+object Optimiser {
+  val log: Logger = LoggerFactory.getLogger(getClass)
+
+  val weightSla = 10
+  val weightChurn = 50
+  val weightPax = 0.05
+  val weightStaff = 3
+  val blockSize = 5
+  val targetWidth = 60
+  val rollingBuffer = 120
+
+  def crunch(workloads: Seq[Double],
+             minDesks: Seq[Int],
+             maxDesks: Seq[Int],
+             config: OptimizerConfig): Try[OptimizerCrunchResult] = {
+    val indexedWork = workloads.toIndexedSeq
+    val indexedMinDesks = minDesks.toIndexedSeq
+
+    val fairMaxD = if (workloads.length >= 60) {
+      val fairXmax = rollingFairXmax(indexedWork, indexedMinDesks, blockSize, (0.75 * config.sla).round.toInt, targetWidth, rollingBuffer)
+      fairXmax.zip(maxDesks).map { case (fair, orig) => List(fair, orig).min }
+    } else maxDesks.toIndexedSeq
+
+    val desks = optimiseWin(indexedWork, indexedMinDesks, fairMaxD, config.sla, weightChurn, weightPax, weightStaff, weightSla)
+    val waits = processWork(indexedWork, desks, config.sla, IndexedSeq()).waits
+
+    Success(OptimizerCrunchResult(desks.toIndexedSeq, waits))
```

Comment:
We can probably get rid of the Try here and just return an OptimizerCrunchResult

Reply:
Maybe after we've got rid of the R optimiser. For the moment we need to the interface to be consistent so they can be swapped in and out